accepted/0000-pure-mode.md

```diff
@@ -0,0 +1,344 @@
+
+Several build tools took advantage of these declared dependencies to split the builds into multiple steps (one per workspace) which could be cached or run in parallel. This caching and parallelism usually improves significantly build performance.
+
+The current default installation strategy does not properly communicates to Node.js the information about the dependency graph. Node.js having an inaccurate view of the dependency graph leads the build systems to wrongly skip build operations, leading to risk of bugs slipping through.
+
+
+The installation mode proposed by this RFC would communicate accurately to Node.js the dependency graph, allowing build tools to correctly reason about it.
+
+## Naming
+
+The proposed name for this mode is `pure-mode`, meaning "pure" as in "no side effects". This name comes from the parallel with functional programming; in functional programming, functions are called `pure` because they don't have side effects, in `pure-mode` workspaces have no side effects.
+
+To make the discussions more efficient, we could also name the current default installation mode of npm. We will call this mode `hoisted-mode` as packages are shared across workspaces by hoisting them to the root of the project.
+
+## More on the problem
+
+Packages and workspaces declare their dependencies to npm by using special fields in their `package.json`. The Node.js runtime is unaware of these special fields, instead it uses a [module resolution algorithm](https://nodejs.org/api/modules.html) to determine the meaning of a module importing another module. npm bridges this gap by converting the declared dependency graph into a folder structure which makes sense to Node.js. This operation is call `reification`.
+
+When converting the dependency graph to a folder structure, `hoisted-mode` loses information. Multiple different dependency graphs can be converted to an identical folder structure. Since the folder structure is all that Node.js uses to resolve modules, Node.js misses some information about dependencies. This loss of information is most frequently visible through the fact that workspaces are able to successfully import the dependencies declared by other workspaces.
+
+### Forgetting to declare dependencies
+
+When workspaces can successfully use code of a package without having a dependency on it, people forget to declare their dependencies. This lead to situations where updating the dependencies of one workspace breaks a seemingly unrelated workspace.
+
+It is worth noting that static code analysis tools can help significantly reduce the frequency of these mistakes.
+
+### Duplication
+
+The `hoisted-mode` is not always successful at sharing common dependencies. Conflicts in versions lead to packages being installed multiple times on disk. These conflicts become more frequent when the project scales. The frequency of these conflicts can be reduced by an effort to align dependencies' versions across the project.
```

Comment:
"sharing" is i think the key word here. I've discussed this concept on the RFC call multiple times; basically, every package.json should have a node_modules next to it, full stop.

However, other sibling packages, and *also* any common peer deps between sibling packages, should be *shared* - without hoisting.

Reply:
Are you suggesting that this new mode should be called `shared-mode`?

Reply:
Yes, and that it should automatically share both duplicate peers as well as child packages.

Reply:
hoisting is also a sharing mechanism so `shared-mode` does not really contrast with hoisting.

Reply:
I know, but this hoisting is sharing 95% of the time. So should we call this new mode "100% sharing-mode"?

Reply:
Hoisting is sharing *all* of them. The new mode I think we need is sharing only peers and siblings, which imo is the only thing everyone actually wants.

That parent package's deps are requireable in children isn't hoisting, it's just a natural artifact of the way node's always worked, and no workspace mode should interfere with that.

Reply:
Making sure that I understand your point well @ljharb :
If packages `foo` depends on `bar` and `cat`, you are saying that `cat` should be able to require `bar` even if no peer dependency to `bar` in declared in `cat`'s package.json?

Reply:
Discussed during the RFC call and @ljharb said that this is not a concern for him anymore.